teuthology/misc.py

```diff
@@ -549,14 +549,15 @@ def sudo_write_file(remote, path, data, perms=None, owner=None):
     owner_args = []
     if owner:
         owner_args = [run.Raw('&&'), 'sudo', 'chown', owner, path]
-    remote.run(
+    return remote.run(
```

Comment:
Instead of this you need to use try-catch and check CommandFailedError.exitstatus value

Reply:
Hmm, agreed. But I would like to say that using try-except is not as elegant as -
```
    retval = remote.run(args=['some', 'cmd']).returncode
   self.assertEqual(retval, 1)
```

Reply:
I'm sorry, but in my taste this is not super elegant. That's more like c-style coding, I don't mind, but we have OOP language which allows to use try-catch constructions. From other side, why do you need assertEqual which generating corresponding exception anyways, and it mask original exception which makes a step back from usability of issue investigation. Also in case of error remote.run will not return any object by default but will throw an exception which you have to handle. If you don't want exception to be occur, you need to pass special parameter to run:
```
        :param check_status: Whether to raise CommandFailedError on non-zero
                             exit status, and . Defaults to True. All signals
                             and connection loss are made to look like SIGHUP.
```
But this only usable if you want to ignore exceptions.
Third, when we use write_file function, we do not expect there is an error so we need to interrupt program, and either use sudo_write_file or we have some a bug definitely.

Reply:
> I'm sorry, but in my taste this is not super elegant. 

Really? Compared to following -

```
try:
    remote.run(args=['some', 'cmd])
except CommandFailedError as e:
    self.assertEqual(e.existatus, 1)
```

I can use check_status, agreed but I still think passing the object representing the subprocess is better than returning void. For example, access to stdout and stderr can help better error diagnosing/more accurate test sometimes.

Reply:
You do not need to have access to stdout or stderr if you are not going to analyze the errors programatically. For example in case of write_file there is no need to do this, since you are not going to continue execution, you just need to stop the test as fast as possible. All stdout and stderr can be found in the logs by anyone who read the docs.
See, the exit status is printed in output:
```
>>> from teuthology.orchestra.remote import Remote as r ; l=r('localhost')
>>> l.write_file('/tmp/bla/bla', 'blabla')
2020-07-17 17:51:26,461.461 INFO:teuthology.orchestra.remote:Trying to reconnect to host
2020-07-17 17:51:26,686.686 INFO:teuthology.orchestra.run.localhost:> true
2020-07-17 17:51:26,804.804 INFO:teuthology.orchestra.run.localhost:> set -ex
2020-07-17 17:51:26,804.804 INFO:teuthology.orchestra.run.localhost:> dd of=/tmp/bla/bla
2020-07-17 17:51:26,855.855 INFO:teuthology.orchestra.run.localhost.stderr:+ dd of=/tmp/bla/bla
2020-07-17 17:51:26,856.856 INFO:teuthology.orchestra.run.localhost.stderr:dd: failed to open '/tmp/bla/bla': No such file or directory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kyr/kshtsk/teuthology/teuthology/orchestra/remote.py", line 481, in write_file
    self.run(args=args, stdin=data)
  File "/home/kyr/kshtsk/teuthology/teuthology/orchestra/remote.py", line 204, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/kyr/kshtsk/teuthology/teuthology/orchestra/run.py", line 446, in run
    r.wait()
  File "/home/kyr/kshtsk/teuthology/teuthology/orchestra/run.py", line 160, in wait
    self._raise_for_status()
  File "/home/kyr/kshtsk/teuthology/teuthology/orchestra/run.py", line 182, in _raise_for_status
    node=self.hostname, label=self.label
teuthology.exceptions.CommandFailedError: Command failed on localhost with status 1: 'set -ex\ndd of=/tmp/bla/bla'
```
The write_file should not return subprocess, it is a blackbox, I can agree that CommandFailedError could include the stderr and stdout in some cases but this not optimal approach because stdout and stderr in general can be very very big, especially for long running processes. Probably this is why it is not included as a member to CommandFailedError. But when it comes for writing file we do not need so much info, since we usually want to be written always, and if not, then stop.

From your code:
```
except CommandFailedError as e:
    self.assertEqual(e.existatus, 1)
```
I don't understand why do you need to ignore error with status not equal to 1? If exitstatus is Zero the exception is not even raised.
So, writing:
```
remote.run(args=['some', 'cmd'])
```
is more elegant than:
```
 retval = remote.run(args=['some', 'cmd']).returncode
   self.assertEqual(retval, 1)
```
You just don't need to call assertEqual and do you need to call assertEqual at all?

Reply:
I'm not sure I understand which goal are you trying to reach.